core/src/Kokkos_Half.hpp

```diff
@@ -47,12 +47,882 @@
 
 #include <type_traits>
 #include <Kokkos_Macros.hpp>
+#include <iosfwd>  // istream & ostream for extraction and insertion ops
+#include <string>
 
-// Include special backend specific versions here
-#include <Cuda/Kokkos_Cuda_Half.hpp>
+////////////////////////////////////////////////////////////////////////////////
+///////// Include special backend specific headers and impl types here /////////
+#include <Cuda/Kokkos_Cuda_Half_Impl_Type.hpp>
+#include <SYCL/Kokkos_SYCL_Half_Impl_Type.hpp>
+////////////////////////////////////////////////////////////////////////////////
+
+#ifdef KOKKOS_IMPL_HALF_TYPE_DEFINED
+
+// KOKKOS_HALF_IS_STORAGE_ONLY: A macro to select which floating_pointer_wrapper
+// operator paths should be used.
+// For CUDA, let the compiler conditionally select when device ops are used
+// For SYCL, Sycl/Kokkos_Sycl_Half_Impl_type.hpp defines
+// KOKKOS_SYCL_HALF_IS_STORAGE_ONLY
+#if defined(__CUDA_ARCH__) || defined(KOKKOS_SYCL_HALF_IS_STORAGE_ONLY)
+#define KOKKOS_HALF_IS_STORAGE_ONLY
+#endif
```

Comment:
I think it should be the other way around. The half type is storage only for CUDA+Host.
```suggestion
#if defined(KOKKOS_ENABLE_CUDA) && !defined(__CUDA_ARCH__)
#define KOKKOS_HALF_IS_STORAGE_ONLY
#endif
```
but requires you to flip all behavior below, otherwise
```suggestion
#if defined(__CUDA_ARCH__) || defined(KOKKOS_ENABLE_SYCL)
#define KOKKOS_HALF_IS_COMPUTE_TYPE
#endif
```
I would prefer the former. In any case, I don't see a reason to define an extra macro for `SYCL`.